### PR TITLE
Feat/mysql search tool

### PR DESCRIPTION
**Pull request by dajianguo**
# Description

a new tool for execute mysql sql or get the table struct

Fixes # (issue)

## Type of Change

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Yes,I have test.

- [ ] TODO

# Suggested Checklist:

- [√] I have performed a self-review of my own code
- [√ ] I have commented my code, particularly in hard-to-understand areas
- [√ ] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/core/tools/provider/builtin/dbutil/dbutil.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from core.tools.provider.builtin_tool_provider import BuiltinToolProviderController
+
+
+class WikiPediaProvider(BuiltinToolProviderController):
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/dbutil/dbutil.py]: Fix the incorrect tool provider name.

File: api/core/tools/provider/builtin/dbutil/dbutil.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  author: Nut # 作者
+  name: dbutil # 名称，唯一，不允许和其他供应商重名
+  label: # 标签，用于前端展示
+    en_US: DbUtil # 英文标签
+    zh_Hans: DbUtil # 中文标签
+  description: # 描述，用于前端展示
+    en_US: DbUtil # 英文描述
+    zh_Hans: DbUtil # 中文描述
+  icon: icon.svg # 图标，需要放置在当前模块的_assets文件夹下
+  tags: # 标签，用于前端展示
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/dbutil/dbutil.yaml]: Remove all the comments.

**Review by bowenliang123 — CHANGES_REQUESTED** [review]
Hi, please make sure all the codes and configs are provided in English for general usage. Including the YAML configs, text messages sent, exception messages and comments.

**Comment by dajianguo** [comment]
sorry, i wil be modify this error

File: api/core/tools/provider/builtin/dbutil/tools/sqlutil.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+from typing import Any, Union
+import logging
+import datetime
+
+from core.tools.entities.tool_entities import ToolInvokeMessage
+from core.tools.tool.builtin_tool import BuiltinTool
+from core.tools.provider.builtin.dbutil.tools.use_db import use_db
+from pandas import Timestamp
+
+class SqlutilTool(BuiltinTool):
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/sqlutil.py]: It should be SqlUtilTool

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
+from sqlalchemy import create_engine, text
+import pandas as pd
+import logging
+
+
+class use_db():
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: You should use `Camel case` to define a class.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    def __init__(self, connection_string: str, logger: logging) -> None:
+        self.connection_string = connection_string
+        self.logger = logger
+        self.engine = create_engine(self.connection_string, pool_size=100,pool_recycle=36)
+
+    # def get_engine(self):
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+    def run_query(self, query_string: str):
+        '''
+        Run SQL Qurey
+        '''
+        result = None
+        if query_string == '':
+            return None
+        try:
+            # engine = self.get_engine()
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        for f in fields:
+            result += f[1]+', '
+
+        return result
+
+    def get_database_list(self):
+        '''
+        Only use this function to get the root databases.
+        '''
+        # engine = self.get_engine()
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        results = cursor.fetchall()
+        return [
+            d[0]
+            for d in results
+            if d[0] not in ["information_schema", "performance_schema", "sys", "mysql"]
+        ]
+
+    def get_fields(self, table_name):
+        """Get column fields about specified table."""
+        # engine = self.get_engine()
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                    table_name
+                )
+            )
+        )
+        fields = cursor.fetchall()
+        return [[field[0], field[1], field[2]] for field in fields]
+
+    def get_table_comments(self, db_name):
+        """get table's comment"""
+        # engine = self.get_engine()
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        table_comments = cursor.fetchall()
+        names = []
+        comments = []
+        for table_comment in table_comments:
+            names.append(table_comment[0])
+            comments.append(table_comment[1])
+
+        return names, comments
+
+        # return [
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        #     (table_comment[0], table_comment[1]) for table_comment in table_comments
+        # ]
+
+    def get_create_table(self, table_name):
+        """get table schema"""
+        # engine = self.get_engine()
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Remove codes if they are not needed.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        """get table in clumns,return a dict"""
+        cursor = self.engine.connect().execute(
+            text(
+                f"SELECT COLUMN_NAME, COLUMN_COMMENT  from information_schema.COLUMNS where table_name='{table_name}'".format(
+                    table_name
+                )
+            )
+        )
+        fields = cursor.fetchall()
+        print(fields)
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Do not use print in the code.

File: api/core/tools/provider/builtin/dbutil/tools/use_db.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        dic: dict = {}
+        for field in fields:
+            print(field)
+            print(field[0],"-----------",field[1])
+
+            dic[field[0]]=str(field[1])
+        print("use_db.get_columns_dict.dic---------------------------------",dic)
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/use_db.py]: Do not use print in the code.

File: api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+class TablestructutilTool(BuiltinTool):
+    def _invoke(
+        self, user_id: str, tool_parameters: dict[str, Any]
+    ) -> Union[ToolInvokeMessage, list[ToolInvokeMessage]]:
+        """
+        invoke tools
+        """
+        db_type = tool_parameters.get("db_type", "")
+        if not db_type:
+            return self.create_text_message("请选择数据库类型")
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py]: All the messages should be translated or in English.

File: api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        db_connection = tool_parameters.get("db_connection", "")
+        if not db_connection:
+            return self.create_text_message("请填写数据库链接信息，例如：mysql+pymysql://admin:admin@localhost:13306/test")
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py]: All the messages should be translated or in English.

File: api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        query_tbl = tool_parameters.get("query_tbl", "")
+        if not query_tbl:
+            return self.create_text_message("请填写表名")
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py]: All the messages should be translated or in English.

File: api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+        try:
+            db = use_db(db_connection, logging)
+
+        except Exception as e:
+            return self.create_text_message(
+                "创建数据库链接异常. {}".format(e)
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py]: All the messages should be translated or in English.

File: api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+            if db_struct == None:
+                db_struct = []
+
+
+            return self.create_text_message(
+                json.dumps(db_struct, ensure_ascii=False)
+            )
+        except Exception as e:
+            return self.create_text_message(
+                "执行sql异常. {}".format(e)
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/tools/tablestructutil.py]: All the messages should be translated or in English.

File: api/core/tools/provider/builtin/dbutil/_assets/icon.svg
Comment (crazywoola) [inline api/core/tools/provider/builtin/dbutil/_assets/icon.svg:1]: This is an icon from another tool right? Please choose a different one.

**Review by crazywoola — CHANGES_REQUESTED** [review]
Please resolve comments :) And thanks for your contribution.

**Comment by dajianguo** [comment]
I have corrected the review comments

**Comment by crazywoola** [comment]
Please fix the CI as well

**Comment by crazywoola** [comment]
Close for now.

**Pull request closed without merging**